truenas_installer/install.py

```diff
@@ -0,0 +1,167 @@
+        finally:
+            await run(["zpool", "export", "-f", BOOT_POOL])
+    except subprocess.CalledProcessError as e:
+        raise InstallError(f"Command {' '.join(e.cmd)} failed:\n{e.stderr.rstrip()}")
+
+
+async def format_disk(device, create_swap, set_pmbr, callback):
+    if (result := await run(["wipefs", "-a", device], check=False)).returncode != 0:
+        callback(0, f"Warning: unable to wipe partition table for {device}: {result.stderr.rstrip()}")
+
+    # Erase both typical metadata area.
+    await run(["sgdisk", "-Z", device], check=False)
+    await run(["sgdisk", "-Z", device], check=False)
+
+    # Create BIOS boot partition
+    await run(["sgdisk", "-a4096", "-n1:0:+1024K", "-t1:EF02", "-A1:set:2", device])
+
+    # Create EFI partition (Even if not used, allows user to switch to UEFI later)
+    await run(["sgdisk", "-n2:0:+524288K", "-t2:EF00", device])
+
+    if create_swap:
+        await run(["sgdisk", "-n4:0:+16777216K", "-t4:8200", device])
+        await run(["wipefs", "-a", "-t", "zfs_member", get_partition(device, 4)], check=False)
+
+    # Create data partition
+    await run(["sgdisk", "-n3:0:0", "-t3:BF01", device])
+
+    # Bad hardware is bad, but we've seen a few users
+    # state that by the time we run `parted` command
+    # down below OR the caller of this function tries
+    # to do something with the partition(s), they won't
+    # be present. This is almost _exclusively_ related
+    # to bad hardware, but we add this here as a compromise.
+    await wait_on_partitions(device, [1, 2, 3, 4] if create_swap else [1, 2, 3])
+
+    if set_pmbr:
+        await run(["parted", "-s", device, "disk_set", "pmbr_boot", "on"], check=False)
+
+
+async def wait_on_partitions(device, partitions):
+    if not pathlib.Path(device).is_block_device():
+        raise InstallError(f"{device} was not found or is not a block device")
+
+    partitions = [get_partition(device, partition) for partition in partitions]
+
+    for i in range(30):
+        if all(pathlib.Path(partition).is_block_device() for partition in partitions):
+            return
+
+        await asyncio.sleep(1)
+
+    for partition in partitions:
+        if not pathlib.Path(partition).is_block_device():
+            raise InstallError(f"Could not find {partition}")
+
+
+async def create_boot_pool(devices):
+    await run(
+        [
+            "zpool", "create", "-f",
+            "-o", "ashift=12",
+            "-o", "cachefile=/tmp/zpool.cache",
+            "-o", "compatibility=grub2",
+            "-O", "acltype=off",
+            "-O", "canmount=off",
+            "-O", "compression=lz4",
+            "-O", "devices=off",
+            "-O", "mountpoint=none",
+            "-O", "normalization=formD",
+            "-O", "relatime=on",
+            "-O", "xattr=sa",
+            BOOT_POOL,
+        ] +
+        (["mirror"] if len(devices) > 1 else []) +
+        devices
+    )
+    await run(["zfs", "set", "compression=on", BOOT_POOL])
```

Comment:
Why do we set `compression=on` after creating the zpool with `compression=lz4`? Seems superfluous to do that here.